File: cpp/api/src/base/WaitingSet.cpp
/*
 * CAMEO
 *
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under BSD 3-Clause and GPL-v3 as described in license files.
 * You may not use this work except in compliance with the Licences.
 *
 */

#include "WaitingSet.h"
#include "Waiting.h"

using namespace std;

namespace cameo {

WaitingSet::WaitingSet() {

}

void WaitingSet::add(Waiting * waiting) {

	lock_guard<mutex> lock {m_mutex};

	m_set.insert(waiting);
}

void WaitingSet::remove(Waiting * waiting) {

	lock_guard<mutex> lock {m_mutex};

	set<Waiting *>::iterator it = m_set.find(waiting);

	if (it != m_set.end()) {
		m_set.erase(it);
	}
}

void WaitingSet::cancelAll() {

	lock_guard<mutex> lock {m_mutex};

	for (set<Waiting *>::iterator it = m_set.begin(); it != m_set.end(); ++it) {
		(*it)->cancel();
	}

}

}

File: java/api/src/main/java/eu/ill/cameo/api/base/WaitingSet.java
/*
 * CAMEO
 *
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under BSD 3-Clause and GPL-v3 as described in license files.
 * You may not use this work except in compliance with the Licences.
 *
 */

package eu.ill.cameo.api.base;


import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;

/**
 * Class defining a set of waitings.
 */
public class WaitingSet {

	private Set<Waiting> waitingSet = new HashSet<Waiting>();

	/**
	 * Adds a Waiting object.
	 * @param waiting The Waiting object.
	 */
	public void add(Waiting waiting) {
		synchronized (waitingSet) {
			waitingSet.add(waiting);
		}	
	}
	
	/**
	 * Removes a Waiting object.
	 * @param waiting The Waiting object.
	 */
	public void remove(Waiting waiting) {
		synchronized (waitingSet) {
			waitingSet.remove(waiting);
		}	
	}
	
	/**
	 * Cancels all the waitings.
	 */
	public void cancelAll() {

		synchronized (waitingSet) {
			// The iteration must be in the synchronized block.
			Iterator<Waiting> i = waitingSet.iterator();
			while (i.hasNext()) {
				Waiting waiting = i.next();
				waiting.cancel();
			}
		}
	}

	/**
	 * Terminates all the waitings.
	 */
	public void terminateAll() {
		
		// Copy the waitings because terminate removes the Waiting object from waitingSet.
		Set<Waiting> waitingSetCopy = new HashSet<Waiting>();
		
		synchronized (waitingSet) {
			waitingSetCopy.addAll(waitingSet);
		}
			
		// The iteration must be in the synchronized block.
		Iterator<Waiting> i = waitingSetCopy.iterator();
		while (i.hasNext()) {
			i.next().terminate();	
		}	
	}
	
}